compiler/rustc_middle/src/values.rs

```diff
@@ -32,13 +32,23 @@ impl<'tcx> Value<TyCtxt<'tcx>> for ty::SymbolName<'_> {
 }
 
 impl<'tcx> Value<TyCtxt<'tcx>> for ty::Binder<'_, ty::FnSig<'_>> {
-    fn from_cycle_error(tcx: TyCtxt<'tcx>, _: &[QueryInfo]) -> Self {
+    fn from_cycle_error(tcx: TyCtxt<'tcx>, stack: &[QueryInfo]) -> Self {
         let err = tcx.ty_error();
-        // FIXME(compiler-errors): It would be nice if we could get the
-        // query key, so we could at least generate a fn signature that
-        // has the right arity.
+
+        let arity = if let Some(frame) = stack.get(0)
+            && frame.query.name == "fn_sig"
```

Comment:
This isn't possible with a little refactoring so this is fine for this PR but imo this shouldn't compare the string but the `DepKind` enum instead  to guard against query renames.

This isn't possible yet because the `DepKind` isn't part of the query frame but I think that it should be. We do have it available where we [create](https://github.com/rust-lang/rust/blob/11663b1b4857ffeafbd85a9a36c234d117373b76/compiler/rustc_query_impl/src/plumbing.rs) the `QueryStackFrame`.

We can then replace the `name: &'static str` with a `dep_kind: DepKind`, but this does require us to add a generic parameter for the dep kind to `QueryStackFrame` and all of the places its used which isn't great but we can use type aliases in `rustc_middle` to make this not too bad. With that we'd just need a `as_str()` function on `DepKind` which should be easy to add with `stringify!()` where it's generated. This way we could get rid of the string comparison here and maybe also improve things in other places.

@cjgillot @jyn514 what do you think?

Reply:
I agree with all you said. Using `if let`s on `DepKind` is definitely the better way to do that.
Though, it may be better in another PR, so we can approve this one.

Reply:
Yes, I'll open an issue for that.